Scripts/Timer.cs

```diff
@@ -0,0 +1,221 @@
+		/// Will reset the timer anyhow.
+		/// </summary>
+		/// <param name="onComplete">Call back when the timer finished</param>
+		public void Start(Action onComplete = null)
+		{
+			this.onComplete = onComplete;
+			IsPaused = false;
+			Reset();
+			AddToUpdateQueue();
+		}
+
+		/// <summary>
+		/// Start the timer.
+		/// Override the existing duration.
+		/// Will reset the timer anyhow.
+		/// </summary>
+		/// <param name="duration">The new duration of the timer</param>
+		/// <param name="onComplete">Call back when the timer finished</param>
+		public void Start(float duration, Action onComplete = null)
+		{
+			Duration = duration;
+			Start(onComplete);
+		}
+
+		/// <summary>
+		/// Start the timer with loops.
+		/// Will reset the timer anyhow.
+		/// </summary>
+		/// <param name="loopCount">Total loop times. 0 is no loop, -1 is infinite loops.</param>
+		/// <param name="onLoopComplete">Call back on the end of each loops</param>
+		/// <param name="onComplete">Call back when the timer finished</param>
+		public void Start(int loopCount, Action onLoopComplete = null, Action onComplete = null)
+		{
+			Start(Duration, loopCount, onLoopComplete, onComplete);
+		}
+
+		/// <summary>
+		/// Start the timer with loops.
+		/// Override the existing duration.
+		/// Will reset the timer anyhow.
+		/// </summary>
+		/// <param name="duration">The new duration of the timer</param>
+		/// <param name="loopCount">Total loop times. 0 is no loop, -1 is infinite loops.</param>
+		/// <param name="onLoopComplete">Call back on the end of each loops</param>
+		/// <param name="onComplete">Call back when the timer finished</param>
+		public void Start(float duration, int loopCount, Action onLoopComplete = null, Action onComplete = null)
+		{
+			LoopCount = loopCount;
+			this.onLoopComplete = onLoopComplete;
+			Start(duration, onComplete);
+		}
+
+		/// <summary>
+		/// Stops the timer without invoke any onComplete callbacks.
+		/// Will empty all loops too.
+		/// </summary>
+		public void Stop()
+		{
+			TimeRemaining = 0;
+			LoopCount = 0;
+			RemoveFromUpdateQueue();
+		}
+
+		/// <summary>
+		/// Pauses the timer
+		/// </summary>
+		public void Pause()
+		{
+			IsPaused = true;
+			RemoveFromUpdateQueue();
+		}
+
+		/// <summary>
+		/// Unpause the timer
+		/// </summary>
+		public void Resume()
+		{
+			IsPaused = false;
+			AddToUpdateQueue();
+		}
+
+		/// <summary>
+		/// Reset the timer, back to its full duration.
+		/// This function will not affact loop counts.
+		/// In addition, this function will not stop, pause or unpause the timer.
+		/// </summary>
+		public void Reset()
+		{
+			TimeRemaining = Duration;
+		}
+
+		private void AddToUpdateQueue()
+		{
+			if (!isUpdating)
+			{
+				isUpdating = true;
+				monoService.OnUpdate += OnUpdateHandler;
+			}
+		}
+
+		private void RemoveFromUpdateQueue()
+		{
+			isUpdating = false;
+			monoService.OnUpdate -= OnUpdateHandler;
+		}
+
+		private void OnUpdateHandler()
```

Comment:
I think we should swap the cases here.

`
if (!IsRunning) 
{
   // This should not occured since we are not waiting for the next update
  // The timer should have been stopped already
  Stop();	
  return;
}

TimeRemaining -= UseScaledTime ? Time.deltaTime : Time.unscaledDeltaTime;
if (TimeRemaining <= 0f)
{
	TimeRemaining = 0f;
	if (IsLooping)
	{
		// Faster than LoopCount--, stop arguing with this little piece of fancy code
		--LoopCount;
		if (onLoopComplete != null) onLoopComplete();
 		// All loops are finished
		if (LoopCount == 0)
		{
			Stop();
			if (onComplete!= null) onComplete();
		}
		// Start next loop
	        else
	        {
			Reset();
		}
	}
	else
	{
		Stop();
		if (onComplete!= null) onComplete();
	}
}
			
`